gslib/utils/stet_util.py

```diff
@@ -0,0 +1,128 @@
+from __future__ import division
+from __future__ import unicode_literals
+
+import os
+import shutil
+
+from gslib import storage_url
+from gslib.utils import execution_util
+from gslib.utils import temporary_file_util
+
+from boto import config
+
+DEFAULT_STET_CONFIG_PATH = '~/.config/stet.yaml'
+
+
+class StetSubcommandName(object):
+  """Enum class for available STET subcommands."""
+  ENCRYPT = 'encrypt'
+  DECRYPT = 'decrypt'
+
+
+def _get_stet_binary_from_path():
+  """Retrieves STET binary from path if available. Python 2 compatible."""
+  for path_directory in os.getenv('PATH').split(os.path.pathsep):
+    binary_path = os.path.join(path_directory, 'stet')
+    if os.path.exists(binary_path):
+      return binary_path
+
+
+def _stet_transform(subcommand, blob_id, in_file_path, out_file_path, logger):
+  """Runs a STET transform on a file.
+
+  Encrypts for uploads. Decrypts for downloads. Automatically populates
+  flags for the STET binary.
+
+  Args:
+    subcommand (StetSubcommandName): Subcommand to call on STET binary.
+    blob_id (str): Cloud URL that binary uses for validation.
+    in_file_path (str): File to be transformed source.
+    out_file_path (str): Where to write result of transform.
+    logger (logging.Logger): For logging STET binary output.
+
+  Returns:
+    stet_temporary_file_url (StorageUrl): Path to result of STET-transform.
+
+  Raises:
+    OSError: STET binary was not present in gsutil's boto config file.
+  """
+  binary_path = config.get('GSUtil', 'stet_binary_path', None)
+  if not binary_path:
+    binary_path = _get_stet_binary_from_path()
+  if not binary_path:
+    raise KeyError('Could not find STET binary in boto config or PATH.')
+
+  config_path = config.get('GSUtil', 'stet_config_path', None)
+  if not config_path:
+    if os.path.exists(DEFAULT_STET_CONFIG_PATH):
+      config_path = DEFAULT_STET_CONFIG_PATH
+    else:
+      raise KeyError(
+          'Could not find STET config in boto config or at {}'.format(
+              DEFAULT_STET_CONFIG_PATH))
+
+  _, stderr = execution_util.ExecuteExternalCommand([
+      os.path.expanduser(binary_path), subcommand,
+      '--config-file=' + os.path.expanduser(config_path),
+      '--blob-id=' + blob_id, in_file_path, out_file_path
+  ])
+  logger.debug(stderr)
+
+
+def encrypt_upload(source_url, destination_url, logger):
+  """Encrypts a file with STET binary before upload.
+
+  Args:
+    source_url (StorageUrl): Copy source.
+    destination_url (StorageUrl): Copy destination.
+    logger (logging.Logger): For logging STET binary output.
+
+  Returns:
+    stet_temporary_file_url (StorageUrl): Path to STET-encrypted file.
+  """
+  in_file = source_url.object_name
+  out_file = temporary_file_util.GetStetTempFileName(source_url)
+  blob_id = destination_url.url_string
+
+  _stet_transform(StetSubcommandName.ENCRYPT, blob_id, in_file, out_file,
+                  logger)
+  return storage_url.StorageUrlFromString(out_file)
```

Comment:
Do we need to wrap this in storage_url? My guess is that the caller will be only interested in the out_file path?

Reply:
It's more consistent with `copy_helper` to work with StorageUrl.
I'd rather do the conversion here than in `copy_helper`, which is super cluttered.
Also, "StorageUrl in, StorageUrl out" makes some intuitive sense.